### DISCUSSION TITLE
How to parse .capnp schema from std::filesystem::path?

### DISCUSSION THREAD

**Discussion by Tetragramm** [capnproto/capnproto#1858] [Q&A]
I am trying to parse schemas from a folder, which I receive as an `std::filesystem::path`, but I am unable to figure out how to convert that path into something that can be passed into `parseFromDirectory`.
`parseFromDirectory` needs a `kj::ReadableDirectory`, which is virtual, and only instantiated by the `kj::Filesystem` `newDiskReadableDirectory`.  
However, it appears that `kj::Filesystem` can only be created from the current working directory and does not appear to allow changing the directory it represents.
`newDiskReadableDirectory` takes an OsFileHandle, which can't be extracted from the `std::filesystem::path` using standard calls.  

So I'm at a loss.

**Comment by Tetragramm** [comment]
Possible solution, though incredibly unwieldy.
Create an in-memory directory, and read all the capnp files using standard c++ and write them into the in-memory directory.  Then pass that to the `parseFromDirectory` call.  
But this is incredibly awkward, and I'm sure it's not the best way.  Also, I think I'd need to read in all the files before parsing begins to make sure all the imports are available.

**Comment by kentonv** [comment]
There are two options here:

1. Open the directory using plain old `open()` (or `CreateFile()` on Windows), then make a `kj::Directory` from the file descriptor:

```
int fd = open(path, O_RDONLY | O_DIRECTORY);
kj::Own<const kj::ReadableDirectory> dir = kj::newDiskReadableDirectory(fd);
```

2. Use KJ filesystem API all the way:

```
auto fs = kj::newDiskFilesystem();
kj::Path kjPath = fs->getCurrentPath().evalNative(path);
kj::Own<const kj::ReadableDirectory> dir = fs->getRoot().openSubdir(kjPath);
```

In the second way, `getCurrentPath().evalNative(path)` tells KJ to derive the absolute path of the file, based on the path string and the current directory. If `path` starts with `/`, then the current directory doesn't matter, KJ will just parse the path and return it. If it doesn't start with `/`, then KJ prefixes the current directory. In both cases, KJ also evaluates away any instances of `..` (removing the preceding path component). Since either way ends up with an absolute path in the end, you can get use `getRoot().openSubdir()` to open it.

Note that letting KJ parse paths technically does not always produce the same result as having the OS do it. Consider in particular the path `/foo/../bar`, and consider what happens if `foo` is actually a symlink to `baz/qux`. KJ will evaluate the path to `/bar`, which is I think what most people expect. However, the OS will evaluate it to `/baz/bar`, because the OS follows the symlink, and then follows `..` to the parent of the linked directory, not back to the symlink's parent. I think the native OS behavior is surprising and usually not what people want. But if it is what you want, then by all means, use the first option above.

**Reply by Tetragramm** [reply]
Hmm, unfortunately, it doesn't seem to parse the absolute path correctly.
```cpp
auto fs = kj::newDiskFilesystem();
kj::Path kjPath = fs->getCurrentPath().eval(std::filesystem::absolute(f.path()).generic_string());
kj::Own<const kj::ReadableDirectory> dir = fs->getRoot().openSubdir(kjPath);
kj::Path fPath = kjPath.append(file.path().filename().generic_string());
```
`parseFromDirectory(dir, fPath)` Fails with a no such file exception.  Printing the file path, and `kjPath` and `fPath` gives this
```
File Path is /home/<snip>/SocketSystemInterface-1.0/schemas
kjPath is home/<snip>/SocketSystemInterface-1.0/schemas
fPath is home/<snip>/SocketSystemInterface-1.0/schemas/SocketSystemInterface.capnp
```
No difference between `.eval` or `.evalNative` in the results.

**Reply by kentonv** [reply]
When you call `parseFromDirectory`, you should be passing it a path relative to the directory, not an absolute path. Something like:

```
kj::Path fPath = kj::Path({file.path().filename().generic_string()});
```

Otherwise I believe the code you have written will attempt to open the file:

```
home/<snip>/SocketSystemInterface-1.0/schemas/home/<snip>/SocketSystemInterface-1.0/schemas/SocketSystemInterface.capnp
```
